ca_pool.go

```diff
@@ -17,91 +17,231 @@
 package identity
 
 import (
+	"crypto/sha256"
 	"crypto/x509"
+	"encoding/hex"
+	"github.com/pkg/errors"
+	"time"
 )
 
 type CaPool struct {
-	certs []*x509.Certificate
+	roots         map[string]*x509.Certificate
+	intermediates map[string]*x509.Certificate
 }
 
 func NewCaPool(certs []*x509.Certificate) *CaPool {
 	result := &CaPool{
-		certs: certs,
+		roots:         make(map[string]*x509.Certificate),
+		intermediates: make(map[string]*x509.Certificate),
+	}
+
+	for _, cert := range certs {
+		_ = result.AddCa(cert)
 	}
 	return result
 }
 
-func (self *CaPool) isSelfSignedCA(cert *x509.Certificate) bool {
-	return cert.IsCA && cert.CheckSignatureFrom(cert) == nil
+// AddCa adds a CA (root or intermediate) certificate to the current pool. It returns an error if the
+// certificate is not CA.
+func (self *CaPool) AddCa(cert *x509.Certificate) error {
+	if cert == nil {
+		return errors.New("cannot add a nil certificate")
+	}
+	if !cert.IsCA {
+		return errors.New("x509 certificates does not have the CA flag set to true")
+	}
+
+	hash := hashCertSha256(cert)
+
+	if IsRootCa(cert) {
+		self.roots[hash] = cert
```

Comment:
sorta feels like a log message indicating a dedupe happened might be helpful

Reply:
Adding multiple times is idempotent, and this is a low-level library. Logging here isn't the correct play as we don't know how it is being used.

If we were going to make logging a thing, this would return an error on duplicate adds, and the caller would have the choice to log in or ignore it. However, it is idempotent and affects nothing, so I chose not to error, as that is how the old implementation worked. Changing it here would require changing every caller out there.